Fix compiler assert fail on circular type inference error (issue #1334)

jemc opened the pull request:
This PR fixes a compiler assert fail on circular type inference error.

Resolves #1334.

Praetonus commented on src/libponyc/expr/reference.c:
```diff
@@ -675,7 +675,7 @@ bool expr_reference(pass_opt_t* opt, ast_t** astp)
       ast_t* type = ast_type(def);
 
-      if(is_typecheck_error(type))
+      if(is_typecheck_error(type) && (errors_get_count(opt->check.errors) > 0))
```
Is it really the best thing to do? In a code like this

```pony
actor Main
  new create(env: Env) =>
    let x = x.create()
    let y = y.create()
```

only the error for `x` is reported. Wouldn't it be better to remove that line entierely?

jemc replied:
The reason I did this is because the approach of removing the line entirely made the error message for another badpony case a little worse: https://github.com/ponylang/ponyc/pull/1339/files#diff-6adc6b8a17f2bbfe51cbc0151f770ce5R263

However, I've updated the PR to use that approach - for now we can live with those consequences I think, and possibly try to improve that other error message in a separate PR.

SeanTAllen commented:
There's a release underway. Please rebase against master and verify your CHANGELOG entry is correctly appearing in the "unreleased" section.

jemc commented:
This has been updated based on @Praetonus' feedback and rebased to use the latest changelog section.

SeanTAllen commented:
CHANGELOG changed due to 0.7.0 release, this probably needs to be rebased against master.

SeanTAllen commented:
@jemc can you rebase one more time?

@Praetonus ready to sign off on this?

Praetonus commented:
Yes, looks good to me.

jemc commented:
It's now rebased, so I'm merging.

However, I removed the changelog entry from the PR because I'm going to use this as a guinea pig for the automated changelog process.
